src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,245 @@
+    // clang-tidy wants this to return a const iterator
+    CollocationIterator operator++(int) noexcept {  // NOLINT
+      return CollocationIterator(collocation_, index_++);
+    }
+
+    /// retreat the iterator by one position (prefix)
+    CollocationIterator& operator--() noexcept {
+      --index_;
+      return *this;
+    };
+    /// retreat the iterator by one position (prefix)
+    // clang-tidy wants this to return a const iterator
+    CollocationIterator operator--(int) noexcept {  // NOLINT
+      return CollocationIterator(collocation_, index_--);
+    }
+
+    // @{
+    /// (In)Equivalence checks both the object and index for the iterator
+    bool operator==(const CollocationIterator& rhs) noexcept;
+    bool operator!=(const CollocationIterator& rhs) noexcept;
+    // @}
+
+   private:
+    size_t index_;
+    const Collocation<Representation>* collocation_;
+  };
+
+  /// The representation of the block of complex values, which sets the stride
+  /// inside the libsharp type
+  static const ComplexRepresentation complex_representation = Representation;
+
+  /// \brief Generates the libsharp collocation information and stores it in
+  /// `geom_info_`.
+  ///
+  /// \note If you will potentially use the same `l_max` collocation set more
+  /// than once, it is probably better to use the
+  /// `precomputed_spherical_harmonic_collocation` function
+  explicit Collocation(size_t l_max) noexcept;
+
+  /// default constructor required for iterator use
+  ~Collocation() = default;
+  Collocation() = default;
+  Collocation(const Collocation&) = default;
+  Collocation(Collocation&&) = default;
+  Collocation& operator=(Collocation&) = default;
+  Collocation& operator=(Collocation&&) = default;
+
+  /// retrieve the `sharp_geom_info*` stored. This should largely be used only
+  /// for passing to other libsharp functions. Otherwise, access elements
+  /// through iterator or access functions.
+  // clang-tidy seems to think the pointer will implicitly convert to the
+  // libsharp type. It does not.
+  sharp_geom_info* get_sharp_geom_info() const noexcept {
+    return *(geom_info_.get());  // NOLINT
+  }
+
+  /// Retrieve the \f$\theta\f$ value for a given index in a libsharp-compatible
+  /// array
+  double theta(size_t offset) const noexcept;
+  /// Retrieve the \f$\phi\f$ value for a given index in a libsharp-compatible
+  /// array
+  double phi(size_t offset) const noexcept;
+
+  constexpr size_t l_max() const noexcept { return l_max_; }
+
+  /// Compute the number of entries the libsharp-compatible data structure
+  /// should have
+  constexpr size_t size() const noexcept {
+    return (l_max_ + 1) * (2 * l_max_ + 1);
+  }
+
+  // @{
+  /// Get a bidirectional iterator to the start of the grid. `operator*` for
+  /// that iterator gives a
+  /// `std::pair<size_t /*offset*/ std::pair<double /*theta*/, double /*phi*/>`
+  const Collocation<Representation>::CollocationIterator begin() const
```

Comment:
This won't need to return `const` (same with `end`). Could you also add `cbegin` and `cend` and have these call those or vice-versa? STL containers typically have both :)